src/mds/Server.cc

```diff
@@ -297,6 +298,94 @@ class C_MDS_session_finish : public ServerLogContext {
   }
 };
 
+Session* Server::find_session_by_uuid(const std::string& uuid)
+{
+  Session* session = nullptr;
+  for (auto& it : mds->sessionmap.get_sessions()) {
+    auto& metadata = it.second->info.client_metadata;
+
+    auto p = metadata.find("uuid");
+    if (p == metadata.end() || p->second != uuid)
+      continue;
+
+    if (!session) {
+      session = it.second;
+    } else if (!session->reclaiming_from) {
+      assert(it.second->reclaiming_from == session);
+      session = it.second;
+    } else {
+      assert(session->reclaiming_from == it.second);
+    }
+  }
+  return session;
+}
+
+void Server::reclaim_session(Session *session, MClientSession *m)
+{
+  if (!session->is_open() && !session->is_stale()) {
+    dout(10) << "session not open, dropping this req" << dendl;
+    return;
+  }
+
+  MClientSession *reply = new MClientSession(CEPH_SESSION_RECLAIM_REPLY);
+  auto it = m->metadata.find("uuid");
+  if (it == m->metadata.end()) {
+    dout(10) << __func__ << " invalid message (no uuid)" << dendl;
+    reply->metadata["reclaim_errno"] = stringify(-EINVAL);
+    mds->send_message_client(reply, session);
+    return;
+  }
+  const auto& uuid = it->second;
+
+  unsigned flags = 0;
+  it = m->metadata.find("flags");
+  if (it != m->metadata.end())
+    flags = strtoul(it->second.c_str(), nullptr, 0);
+
+  if (flags != CEPH_RECLAIM_RESET) {
+    dout(10) << __func__ << " unsupported flags" << dendl;
+    reply->metadata["reclaim_errno"] = stringify(-EOPNOTSUPP);
+    mds->send_message_client(reply, session);
+    return;
+  }
+
+  Session* target = find_session_by_uuid(uuid);
```

Comment:
We need to log the session is being reclaimed to handle MDS failover otherwise the client requests will fail.

Reply:
Also, what should happen if two clients try to reclaim the same session? Or, one client tries to reclaim and fails then a second one tries? I think we should only allow session recovery once. Is it problematic to have the new session gain all of the caps/locks of the old session immediately? Then we can (a) discard the old session and allow the client to mark which caps/locks are needed. Once reclaim is finished, release the caps/locks not needed.

The important bit is that the client can log its new uuid (e.g. to a recovery database) so that if it fails during reclaim, the next instance can reclaim against the new uuid. **I don't think it's important that the new session adopts the old uuid.**

Reply:
if two clients try to reclaim the same session, mds considers the old reclaimer (not call reclaim finish yes) is dead, the new reclaimer gain all of the caps/locks of the old reclaimer.  (the logic is not in this RP, see https://github.com/ukernel/ceph/tree/wip-cephfs-state-reclaim2)

using new uuid is racy. If previous reclaimer crashed, current reclaimer does not know if previous reclaimer succeeded in reclaiming state. so current reclaimer needs to pass two uuids to mds. if current reclaimer crashes, next reclaimer needs to pass three uuids to mds.

Reply:
> using new uuid is racy. If previous reclaimer crashed, current reclaimer does not know if previous reclaimer succeeded in reclaiming state. so current reclaimer needs to pass two uuids to mds.

It's possible to log this (in the rados recovery db for instance) in a way so that it works but I think I agree now that it's not necessary.